src/tools/GraphemeTableGen/Program.cs

```diff
@@ -278,32 +289,51 @@ static Ucd ExtractValuesFromUcd(string path)
                 "Linker" => ClusterBreak.InCBLinker,
                 "Consonant" => ClusterBreak.InCBConsonant,
-                _ => throw new Exception($"Unrecognized InCB {indicConjunctBreak} for {firstCp} to {lastCp}")
+                _ => throw new Exception($"Unrecognized InCB {indicConjunctBreak} for U+{firstCp:X4} to U+{lastCp:X4}")
             };
 
             var width = eastAsian switch
             {
                 "N" or "Na" or "H" => CharacterWidth.Narrow, // Half-width, Narrow, Neutral
                 "F" or "W" => CharacterWidth.Wide, // Wide, Full-width
                 "A" => CharacterWidth.Ambiguous, // Ambiguous
-                _ => throw new Exception($"Unrecognized ea {eastAsian} for {firstCp} to {lastCp}")
+                _ => throw new Exception($"Unrecognized ea {eastAsian} for U+{firstCp:X4} to U+{lastCp:X4}")
             };
 
             // There's no "ea" attribute for "zero width" so we need to do that ourselves. This matches:
             //   Me: Mark, enclosing
             //   Mn: Mark, non-spacing
             //   Cf: Control, format
-            if (generalCategory == "Me" || generalCategory == "Mn" || generalCategory == "Cf")
+            switch (generalCategory)
             {
-                width = CharacterWidth.ZeroWidth;
+                case "Cf" when cb == ClusterBreak.Control:
+                    // A significant portion of Cf characters are not just gc=Cf (= commonly considered zero-width),
+                    // but also GCB=CN (= does not join). This is a bit of a problem for terminals,
+                    // because they don't support zero-width graphemes, as zero-width columns can't exist.
+                    // So, we turn all of them into Extend, which is roughly how wcswidth() would treat them.
+                    cb = ClusterBreak.Extend;
+                    width = CharacterWidth.ZeroWidth;
+                    break;
+                case "Me" or "Mn" or "Cf":
+                    width = CharacterWidth.ZeroWidth;
+                    break;
             }
 
             Fill(firstCp, lastCp, TrieValue(cb, width));
         }
     }
 
-    // Box-drawing and block elements are ambiguous according to their EastAsian attribute,
-    // but by convention terminals always consider them to be narrow.
+    // U+00AD: Soft Hyphen
+    // A soft hyphen is a hint that a word break is allowed at that position.
+    // By default, the glyph is supposed to be invisible, and only if
+    // a word break occurs, the text renderer should display a hyphen.
+    // A terminal does not support computerized typesetting, but unlike the other
+    // gc=Cf cases we give it a Narrow width, because that matches wcswidth().
+    Fill(0x00AD, 0x00AD, TrieValue(ClusterBreak.Control, CharacterWidth.Narrow));
+
+    // U+2500 to U+257F: Box Drawing block
+    // U+2580 to U+259F: Block Elements block
+    // By default, CharacterWidth.Ambiguous, but by convention .Narrow in terminals.

```

Comment:
oh no. i now realize that conhost may have expected these to be Ambiguous - in applications that target CJK and do box drawing. i can't remember...